src/lib.rs

```diff
@@ -11,28 +11,214 @@ use rand::Rng;
+        if self.bounds != None {
+            if cell.0 <= self.bounds.unwrap().0 && cell.1 <= self.bounds.unwrap().1 {
+                self.grid.insert(cell);
+            }
+        } else {
+            self.grid.insert(cell);
+        }
+    }
+    fn contains(&self, cell: &Cell) -> bool {
+        self.grid.contains(cell)
+    }
+
+    fn x_bound(&self) -> usize {
+        self.bounds.unwrap().0
+    }
+
+    fn y_bound(&self) -> usize {
+        self.bounds.unwrap().1
+    }
+
+    pub fn gen(&mut self) {
+        if self.bounds == None {
+            return;
+        }
+        self.grid.clear();
+        for x in 0..self.bounds.unwrap().0 {
+            for y in 0..self.bounds.unwrap().1 {
+                if rand::thread_rng().gen_weighted_bool(10) {
+                    self.insert((x, y));
+                }
+            }
+        }
+    }
+}
+
+impl From<Vec<String>> for Grid {
+    /// Returns a Grid interpreted from a string representation
+    ///
+    /// # Arguments
+    ///
+    /// * `s` - Representation of the grid. Each element of the vector
+    /// represents a row in the grid. Hash marks # indicate live cells.
+    /// For example ovec!["#  ", "   ", " # "] represents a grid with live
+    /// cells at (0, 0) and (2, 1).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # #[macro_use] extern crate hemoglobin; fn main(){
+    ///
+    /// let grid = hemoglobin::Grid::from(ovec!["#  ", "   ", " # "]);
+    /// # }
+    /// ```
+    fn from(s: Vec<String>) -> Self {
+        let mut result = Grid::new(None);
+        for (y, row) in s.iter().enumerate() {
+            for (x, c) in row.chars().enumerate() {
+                if c == '#' {
+                    result.insert((x, y));
+                }
+            }
+        }
+        result
+    }
+}
+
+pub struct Rule {
+    dec: BigInt,
+    bin: BitVec,
+}
+
+impl From<BigInt> for Rule {
+    fn from(x: BigInt) -> Self {
+        Rule {
+            dec: x.clone(),
+            bin: {
+                let result_reversed = BitVec::from_bytes(&x.to_bytes_be().1);
+                let mut result = BitVec::from_elem(512, false);
+                for i in 0..result_reversed.len() {
+                    result.set(i, result_reversed[result_reversed.len() - i - 1]);
+                }
+                result
+            },
+        }
+    }
+}
+
+impl From<String> for Rule {
+    fn from(s: String) -> Self {
+        Rule::from(s.parse::<BigInt>().unwrap())
+    }
+}
+
+pub struct World {
+    rule: Rule,
+    grid: Grid,
+}
+
+impl World {
+    pub fn new(width: usize, height: usize, rule: Rule) -> Self {
+        World {
+            rule: rule,
+            grid: Grid::new(Some((width, height))),
+        }
+    }
+
+    fn decide_next_state(&self, cell: &Cell) -> bool {
+        let state = get_state(&self.grid, cell);
+        self.rule.bin[state]
+    }
+
+    pub fn step(&mut self) {
+        let mut new_state = Grid::new(self.grid.bounds);
```

Comment:
I think it would be better not to allocate a new grid at each step, but rather to preallocate two grids and "double-buffer" between them. (This probably doesn't make much difference in the case where a grid uses a hashtable, but if the grid state were represented by a `Vec<BitVec>` instead, or something like that, then having a double buffer would mean there's no need to allocate memory at each step.

Reply:
This is true. It also seems like it will require the use of lifetimes. I'm going to study it now and try and fix the other issue that also needs it.